File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/pager/PagerReaderFragment.java
package eu.kanade.tachiyomi.ui.reader.viewer.pager;

import android.graphics.PointF;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.davemorrissey.labs.subscaleview.ImageSource;
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;

import java.io.File;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import butterknife.Bind;
import butterknife.ButterKnife;
import eu.kanade.tachiyomi.R;
import eu.kanade.tachiyomi.data.source.model.Page;
import eu.kanade.tachiyomi.ui.base.fragment.BaseFragment;
import eu.kanade.tachiyomi.ui.reader.ReaderActivity;
import eu.kanade.tachiyomi.ui.reader.viewer.base.PageDecodeErrorLayout;
import eu.kanade.tachiyomi.ui.reader.viewer.pager.horizontal.RightToLeftReader;
import eu.kanade.tachiyomi.ui.reader.viewer.pager.vertical.VerticalReader;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subjects.PublishSubject;

public class PagerReaderFragment extends BaseFragment {

    @Bind(R.id.page_image_view) SubsamplingScaleImageView imageView;
    @Bind(R.id.progress_container) LinearLayout progressContainer;
    @Bind(R.id.progress) ProgressBar progressBar;
    @Bind(R.id.progress_text) TextView progressText;
    @Bind(R.id.retry_button) Button retryButton;

    private Page page;
    private Subscription progressSubscription;
    private Subscription statusSubscription;
    private int position = -1;

    private int lightGreyColor;
    private int blackColor;

    public static PagerReaderFragment newInstance() {
        return new PagerReaderFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.item_pager_reader, container, false);
        ButterKnife.bind(this, view);
        ReaderActivity activity = getReaderActivity();
        PagerReader parentFragment = (PagerReader) getParentFragment();

        lightGreyColor = ContextCompat.getColor(getContext(), R.color.light_grey);
        blackColor = ContextCompat.getColor(getContext(), R.color.primary_text);

        if (activity.getReaderTheme() == ReaderActivity.BLACK_THEME) {
             progressText.setTextColor(lightGreyColor);
        }

        if (parentFragment instanceof RightToLeftReader) {
            view.setRotation(-180);
        }

        imageView.setParallelLoadingEnabled(true);
        imageView.setMaxBitmapDimensions(activity.getMaxBitmapSize());
        imageView.setDoubleTapZoomStyle(SubsamplingScaleImageView.ZOOM_FOCUS_FIXED);
        imageView.setPanLimit(SubsamplingScaleImageView.PAN_LIMIT_INSIDE);
        imageView.setMinimumScaleType(parentFragment.scaleType);
        imageView.setMinimumDpi(50);
        imageView.setRegionDecoderClass(parentFragment.getRegionDecoderClass());
        imageView.setBitmapDecoderClass(parentFragment.getBitmapDecoderClass());
        imageView.setVerticalScrollingParent(parentFragment instanceof VerticalReader);
        imageView.setOnTouchListener((v, motionEvent) -> parentFragment.gestureDetector.onTouchEvent(motionEvent));
        imageView.setOnImageEventListener(new SubsamplingScaleImageView.DefaultOnImageEventListener() {
            @Override
            public void onReady() {
                switch (parentFragment.zoomStart) {
                    case PagerReader.ALIGN_LEFT:
                        imageView.setScaleAndCenter(imageView.getScale(), new PointF(0, 0));
                        break;
                    case PagerReader.ALIGN_RIGHT:
                        imageView.setScaleAndCenter(imageView.getScale(), new PointF(imageView.getSWidth(), 0));
                        break;
                    case PagerReader.ALIGN_CENTER:
                        PointF center = imageView.getCenter();
                        center.y = 0;
                        imageView.setScaleAndCenter(imageView.getScale(), center);
                        break;
                }
            }

            @Override
            public void onImageLoadError(Exception e) {
                showImageDecodeError();
            }
        });

        retryButton.setOnTouchListener((v, event) -> {
            if (event.getAction() == MotionEvent.ACTION_UP) {
                activity.getPresenter().retryPage(page);
            }
            return true;
        });

        observeStatus();
        return view;
    }

    @Override
    public void onDestroyView() {
        unsubscribeProgress();
        unsubscribeStatus();
        imageView.setOnTouchListener(null);
        imageView.setOnImageEventListener(null);
        ButterKnife.unbind(this);
        super.onDestroyView();
    }

    public void setPage(Page page) {
        this.page = page;

        // This method can be called before the view is created
        if (imageView != null) {
            observeStatus();
        }
    }

    public void setPosition(int position) {
        this.position = position;
    }

    private void showImage() {
        if (page == null || page.getImagePath() == null)
            return;

        File imagePath = new File(page.getImagePath());
        if (imagePath.exists()) {
            imageView.setImage(ImageSource.uri(page.getImagePath()));
            progressContainer.setVisibility(View.GONE);
        } else {
            page.setStatus(Page.ERROR);
        }
    }

    private void showDownloading() {
        progressContainer.setVisibility(View.VISIBLE);
        progressText.setVisibility(View.VISIBLE);
    }

    private void showLoading() {
        progressContainer.setVisibility(View.VISIBLE);
        progressText.setVisibility(View.VISIBLE);
        progressText.setText(R.string.downloading);
    }

    private void showError() {
        progressContainer.setVisibility(View.GONE);
        retryButton.setVisibility(View.VISIBLE);
    }

    private void hideError() {
        retryButton.setVisibility(View.GONE);
    }

    private void showImageDecodeError() {
        ViewGroup view = (ViewGroup) getView();
        if (view == null)
            return;

        LinearLayout errorLayout = new PageDecodeErrorLayout(getContext(), page,
                getReaderActivity().getReaderTheme(),
                () -> getReaderActivity().getPresenter().retryPage(page));

        view.addView(errorLayout);
    }

    private void processStatus(int status) {
        switch (status) {
            case Page.QUEUE:
                hideError();
                break;
            case Page.LOAD_PAGE:
                showLoading();
                break;
            case Page.DOWNLOAD_IMAGE:
                observeProgress();
                showDownloading();
                break;
            case Page.READY:
                showImage();
                unsubscribeProgress();
                break;
            case Page.ERROR:
                showError();
                unsubscribeProgress();
                break;
        }
    }

    private void observeStatus() {
        if (page == null || statusSubscription != null)
            return;

        PublishSubject<Integer> statusSubject = PublishSubject.create();
        page.setStatusSubject(statusSubject);

        statusSubscription = statusSubject
                .startWith(page.getStatus())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::processStatus);
    }

    private void observeProgress() {
        if (progressSubscription != null)
            return;

        final AtomicInteger currentValue = new AtomicInteger(-1);

        progressSubscription = Observable.interval(100, TimeUnit.MILLISECONDS, Schedulers.newThread())
                .onBackpressureLatest()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(tick -> {
                    // Refresh UI only if progress change
                    if (page.getProgress() != currentValue.get()) {
                        currentValue.set(page.getProgress());
                        progressText.setText(getString(R.string.download_progress, page.getProgress()));
                    }
                });
    }

    private void unsubscribeStatus() {
        if (statusSubscription != null) {
            page.setStatusSubject(null);
            statusSubscription.unsubscribe();
            statusSubscription = null;
        }
    }

    private void unsubscribeProgress() {
        if (progressSubscription != null) {
            progressSubscription.unsubscribe();
            progressSubscription = null;
        }
    }

    public Page getPage() {
        return page;
    }

    public int getPosition() {
        return position;
    }

    private ReaderActivity getReaderActivity() {
        return (ReaderActivity) getActivity();
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/viewer/pager/PagerReaderFragment.kt
package eu.kanade.tachiyomi.ui.reader.viewer.pager

import android.graphics.PointF
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import com.davemorrissey.labs.subscaleview.ImageSource
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.source.model.Page
import eu.kanade.tachiyomi.ui.base.fragment.BaseFragment
import eu.kanade.tachiyomi.ui.reader.ReaderActivity
import eu.kanade.tachiyomi.ui.reader.viewer.base.PageDecodeErrorLayout
import eu.kanade.tachiyomi.ui.reader.viewer.pager.horizontal.RightToLeftReader
import eu.kanade.tachiyomi.ui.reader.viewer.pager.vertical.VerticalReader
import kotlinx.android.synthetic.main.chapter_image.*
import kotlinx.android.synthetic.main.item_pager_reader.*
import rx.Observable
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subjects.PublishSubject
import java.io.File
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * Fragment for a single page of the ViewPager reader.
 * All the elements from the layout file "item_pager_reader" are available in this class.
 */
class PagerReaderFragment : BaseFragment() {

    companion object {
        /**
         * Creates a new instance of this fragment.
         *
         * @return a new instance of [PagerReaderFragment].
         */
        fun newInstance(): PagerReaderFragment {
            return PagerReaderFragment()
        }
    }

    /**
     * Page of a chapter.
     */
    var page: Page? = null
        set(value) {
            field = value
            // Observe status if the view is initialized
            if (view != null) {
                observeStatus()
            }
        }

    /**
     * Position of the fragment in the adapter.
     */
    var position = -1

    /**
     * Subscription for progress changes of the page.
     */
    private var progressSubscription: Subscription? = null

    /**
     * Subscription for status changes of the page.
     */
    private var statusSubscription: Subscription? = null

    /**
     * Text color for black theme.
     */
    private val whiteColor by lazy { ContextCompat.getColor(context, R.color.textColorSecondaryDark) }

    /**
     * Text color for white theme.
     */
    private val blackColor by lazy { ContextCompat.getColor(context, R.color.textColorSecondaryLight) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?): View? {
        return inflater.inflate(R.layout.item_pager_reader, container, false)
    }

    override fun onViewCreated(view: View, savedState: Bundle?) {
        if (readerActivity.readerTheme == ReaderActivity.BLACK_THEME) {
            progress_text.setTextColor(whiteColor)
        } else {
            progress_text.setTextColor(blackColor)
        }

        if (pagerReader is RightToLeftReader) {
            view.rotation = -180f
        }

        with(image_view) {
            setParallelLoadingEnabled(true)
            setMaxBitmapDimensions(readerActivity.maxBitmapSize)
            setDoubleTapZoomStyle(SubsamplingScaleImageView.ZOOM_FOCUS_FIXED)
            setPanLimit(SubsamplingScaleImageView.PAN_LIMIT_INSIDE)
            setMinimumScaleType(pagerReader.scaleType)
            setMinimumDpi(50)
            setRegionDecoderClass(pagerReader.regionDecoderClass)
            setBitmapDecoderClass(pagerReader.bitmapDecoderClass)
            setVerticalScrollingParent(pagerReader is VerticalReader)
            setOnTouchListener { v, motionEvent -> pagerReader.gestureDetector.onTouchEvent(motionEvent) }
            setOnImageEventListener(object : SubsamplingScaleImageView.DefaultOnImageEventListener() {
                override fun onReady() {
                    when (pagerReader.zoomType) {
                        PagerReader.ALIGN_LEFT -> setScaleAndCenter(scale, PointF(0f, 0f))
                        PagerReader.ALIGN_RIGHT -> setScaleAndCenter(scale, PointF(sWidth.toFloat(), 0f))
                        PagerReader.ALIGN_CENTER -> {
                            val newCenter = center
                            newCenter.y = 0f
                            setScaleAndCenter(scale, newCenter)
                        }
                    }
                }

                override fun onImageLoadError(e: Exception) {
                    onImageDecodeError()
                }
            })
        }

        retry_button.setOnTouchListener { v, event ->
            if (event.action == MotionEvent.ACTION_UP) {
                readerActivity.presenter.retryPage(page)
            }
            true
        }

        observeStatus()
    }

    override fun onDestroyView() {
        unsubscribeProgress()
        unsubscribeStatus()
        image_view.setOnTouchListener(null)
        image_view.setOnImageEventListener(null)
        super.onDestroyView()
    }

    /**
     * Observes the status of the page and notify the changes.
     *
     * @see processStatus
     */
    private fun observeStatus() {
        page?.let { page ->
            val statusSubject = PublishSubject.create<Int>()
            page.setStatusSubject(statusSubject)

            statusSubscription?.unsubscribe()
            statusSubscription = statusSubject.startWith(page.status)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe { processStatus(it) }
        }
    }

    /**
     * Observes the progress of the page and updates view.
     */
    private fun observeProgress() {
        val currentValue = AtomicInteger(-1)

        progressSubscription?.unsubscribe()
        progressSubscription = Observable.interval(100, TimeUnit.MILLISECONDS, Schedulers.newThread())
                .onBackpressureLatest()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    // Refresh UI only if progress change
                    if (page?.progress != currentValue.get()) {
                        currentValue.set(page?.progress ?: 0)
                        progress_text.text = getString(R.string.download_progress, currentValue.get())
                    }
                }
    }

    /**
     * Called when the status of the page changes.
     *
     * @param status the new status of the page.
     */
    private fun processStatus(status: Int) {
        when (status) {
            Page.QUEUE -> hideError()
            Page.LOAD_PAGE -> onLoading()
            Page.DOWNLOAD_IMAGE -> {
                observeProgress()
                onDownloading()
            }
            Page.READY -> {
                onReady()
                unsubscribeProgress()
            }
            Page.ERROR -> {
                onError()
                unsubscribeProgress()
            }
        }
    }

    /**
     * Unsubscribes from the status subscription.
     */
    private fun unsubscribeStatus() {
        page?.setStatusSubject(null)
        statusSubscription?.unsubscribe()
        statusSubscription = null
    }

    /**
     * Unsubscribes from the progress subscription.
     */
    private fun unsubscribeProgress() {
        progressSubscription?.unsubscribe()
        progressSubscription = null
    }

    /**
     * Called when the page is loading.
     */
    private fun onLoading() {
        progress_container.visibility = View.VISIBLE
        progress_text.visibility = View.VISIBLE
        progress_text.setText(R.string.downloading)
    }

    /**
     * Called when the page is downloading.
     */
    private fun onDownloading() {
        progress_container.visibility = View.VISIBLE
        progress_text.visibility = View.VISIBLE
    }

    /**
     * Called when the page is ready.
     */
    private fun onReady() {
        page?.imagePath?.let { path ->
            if (File(path).exists()) {
                image_view.setImage(ImageSource.uri(path))
                progress_container.visibility = View.GONE
            } else {
                page?.status = Page.ERROR
            }
        }
    }

    /**
     * Called when the page has an error.
     */
    private fun onError() {
        progress_container.visibility = View.GONE
        retry_button.visibility = View.VISIBLE
    }

    /**
     * Hides the error layout.
     */
    private fun hideError() {
        retry_button.visibility = View.GONE
    }

    /**
     * Called when an image fails to decode.
     */
    private fun onImageDecodeError() {
        val view = view as? ViewGroup ?: return

        page?.let { page ->
            val errorLayout = PageDecodeErrorLayout(context, page, readerActivity.readerTheme,
                    { readerActivity.presenter.retryPage(page) })

            view.addView(errorLayout)
        }
    }

    /**
     * Property to get the reader activity.
     */
    private val readerActivity: ReaderActivity
        get() = activity as ReaderActivity

    /**
     * Property to get the pager reader.
     */
    private val pagerReader: PagerReader
        get() = parentFragment as PagerReader

}
